File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N] = input[0].split(" ").map(Number);
const [M] = input[1].split(" ").map(Number);
//const arr = input[1].split("");
const map = input.slice(2, M + 2).map((_) => _.trim().split(" ").map(Number));
const [start, end] = input[M + 2].split(" ").map(Number);

function main() {
  const graph = Array.from({ length: N + 1 }, () =>
    new Array(N + 1).fill(Infinity)
  );

  map.forEach(([start, end, cost]) => {
    graph[start][end] = Math.min(graph[start][end], cost);
  });

  for (let i = 1; i <= N; i++) {
    graph[i][i] = 0;
  }

  const dist = graph[start];
  const visited = Array.from({ length: N + 1 }, () => false);

  visited[start] = true;

  const smallValueNotVisited = () => {
    let min = Infinity,
      minPos = 0;

    for (let i = 1; i <= N; i++) {
      if (dist[i] < min && !visited[i]) {
        min = dist[i];
        minPos = i;
      }
    }

    return minPos;
  };

  for (let i = 1; i < N; i++) {
    const node = smallValueNotVisited();
    visited[node] = true;
    for (let j = 1; j <= N; j++) {
      if (!visited[j]) {
        if (dist[node] + graph[node][j] < dist[j]) {
          dist[j] = dist[node] + graph[node][j];
        }
      }
    }
  }

  console.log(dist[end]);
}

main();

File: 백준/Gold/1916. 최소비용 구하기/최소비용 구하기.cc
#include <iostream>
#include <vector>
using namespace std;

int N, M;
int graph[1001][1001];
bool visited[1001];
int dist[1001];

int smallCostIndex() {
    int min = 987654321;
    int minPos = 0;
    for(int i=1; i<=N; i++) {
        if(dist[i] < min && !visited[i]) {
            min = dist[i];
            minPos = i;
        }
    }
    return minPos;
   
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    //freopen("input.txt", "r", stdin);

    cin >> N >> M;
    fill(&graph[0][0], &graph[N][N+1], 987654321);

    for(int i=0; i<M; i++) {
        int start, end, cost;
        cin >> start >> end >> cost;
        graph[start][end] = min(graph[start][end], cost);
    }

    
    for(int i=1; i<=N; i++) {
        graph[i][i] = 0;
    }

    int start, end;
    cin >> start >> end;

    for(int i=1;i<=N;i++) {
        dist[i] = graph[start][i];
    }

    fill_n(visited, N+1, false);
    visited[start] = true;

    for(int i = 1; i < N; i++) {
        int node = smallCostIndex();
        visited[node] = true;

        for(int j=1; j<=N; j++) {
            if(!visited[j]) {
                if(dist[node] + graph[node][j] < dist[j]) {
                    dist[j] = dist[node] + graph[node][j];
                }
            }
        }
    }

    cout << dist[end];


    return 0;
}